c++/src/kj/filesystem-disk-unix.c++

```diff
@@ -383,7 +383,7 @@ public:
   }
 
   void zero(uint64_t offset, uint64_t size) const {
-#ifdef FALLOC_FL_PUNCH_HOLE
+#if (FALLOC_FL_PUNCH_HOLE && (!(defined(__ANDROID__) && __ANDROID_API__ < 21)))
```

Comment:
```suggestion
// If FALLOC_FL_PUNCH_HOLE is defined, use it to efficiently zero the area.
//
// A fallocate() wrapper was only added to Android's Bionic C library as of API level 21,
// but FALLOC_FL_PUNCH_HOLE is defined in the Linux headers before that, so we'll
// have to explicitly test for that case.
#if defined(FALLOC_FL_PUNCH_HOLE) && !(__ANDROID__ && __BIONIC__ && __ANDROID_API__ < 21)
```

Reply:
I suppose in theory we could probably use `syscall(__NR_fallocate)` since the kernel supports it, but perhaps that's trying too hard, heh.

Reply:
In case of Android would not rely on Linux syscalls not in wide use there. Since the kernel there is only a reference thing for end-user hardware producers, there always be some variant of kernel with a particular thing messed up by their patches.